### PR TITLE
Adding multipart post capability to the Net::HTTP adapter

### PR DESCRIPTION
**Pull request by keeguon**
I added support for the multipart post using the following gem as a dependency: https://github.com/nicksieger/multipart-post as it might be useful to some people.

### CODE REVIEW

**Comment by rogerleite** [comment]
Hi @keeguon!

Thanks for the PR, unfortunately, we're avoiding to add new dependencies to HTTPI. 😐 

In order to use [multipart-post](https://github.com/nicksieger/multipart-post) features, we could have an adapter, like we already have with `net_http_persistent` from [HTTPI::Adapter](https://github.com/savonrb/httpi/blob/master/lib/httpi/adapter.rb#L16). This way we can use multipart-post as optional dependency. What do you think @keeguon?

If you liked the idea, you can create a `net_http_multipart` adapter. Use [HTTPI::Adapter:: NetHTTPPersistent](https://github.com/savonrb/httpi/blob/master/lib/httpi/adapter/net_http_persistent.rb) and it's specs as reference.

Cheers,
Roger Leite

**Comment by keeguon** [comment]
@rogerleite okay yeah this sounds good to me, I'll move my code in a new adapter later today.

File: httpi.gemspec [hunk 1]
```diff
@@ -18,6 +18,7 @@ Gem::Specification.new do |s|
 
   s.add_dependency 'rack'
   s.add_dependency 'socksify'
+  s.add_dependency 'multipart-post'
```
Comment (rogerleite) [inline httpi.gemspec]: Please change to `s.add_development_dependency`, this gem should be "optional" to httpi.

File: httpi.gemspec [hunk 2]
```diff
@@ -25,6 +25,7 @@ Gem::Specification.new do |s|
   s.add_development_dependency 'mocha',    '~> 0.13'
   s.add_development_dependency 'puma',     '~> 2.3.2'
   s.add_development_dependency 'webmock'
+  s.add_development_dependency 'multipart-post'
```
Comment (rogerleite) [inline httpi.gemspec]: Hi @keeguon! Sorry! The "http gems" should be on `Gemfile`, i didn't know too! 😳

File: lib/httpi/adapter.rb [hunk 3]
```diff
@@ -13,7 +13,7 @@ module Adapter
     ADAPTERS = {}
     ADAPTER_CLASS_MAP = {}
 
-    LOAD_ORDER = [:httpclient, :curb, :em_http, :excon, :http, :net_http, :net_http_persistent]
+    LOAD_ORDER = [:httpclient, :curb, :em_http, :excon, :http, :net_http, :net_http_persistent, :net_http_multipart]
```
Comment (rogerleite) [inline lib/httpi/adapter.rb]: This `LOAD_ORDER` is actually wrong, because always stop at `:net_http` symbol. In the future this should be deprecated.

Could you change to this `LOAD_ORDER = [:httpclient, :curb, :em_http, :excon, :http, :net_http]`?

Thanks!

**Comment by rogerleite** [comment]
Hi again @keeguon! Thanks for the hard work here 👍 

This new adapter needs a spec. Can be something like `https://github.com/savonrb/httpi/blob/master/spec/httpi/adapter/net_http_persistent_spec.rb`, to test all "casual" scenarios **and** a scenario of attachment too.

After this spec, it'd be great to squash all commits and will be good to merge.
Send a comment if you need help, i'll try to help.

**Comment by keeguon** [comment]
@rogerleite I'll write the specs later today, testing it out on a modified version of Savon 2 since the final goal here would be to add support for attachments in SOAP requests. Will squash all the commits when it is done ;).

**Comment by keeguon** [comment]
@rogerleite I added specs for the new adapter which led me to modify some stuff in the specs and also I had to modify this line: https://github.com/savonrb/httpi/blob/master/lib/httpi/adapter/net_http.rb#L39 the main `net/http` adapter because overwriting the `body` seems weird and all hell breaks loose if you use this: http://ruby-doc.org/stdlib-2.3.1/libdoc/net/http/rdoc/Net/HTTPGenericRequest.html#method-i-body_stream-3D which in the case of a multipart POST/PUT is mandatory. If you find some other issues regarding this update tell me, I'll do my best to comply.

File: Gemfile [hunk 4]
```diff
@@ -10,6 +10,7 @@ gem 'em-http-request',                  :require => false, :platforms => [:ruby,
 gem 'em-synchrony',                     :require => false, :platforms => [:ruby, :jruby]
 gem 'excon',               '~> 0.21',   :require => false, :platforms => [:ruby, :jruby]
 gem 'net-http-persistent', '~> 2.8',    :require => false
+gem 'multipart-post',                   :require => false
```
Comment (rogerleite) [inline Gemfile]: Hi @keeguon. Looking again all changes, i realized that is missing the second argument and maybe this is a problem to jruby. You can try to change to:

`gem 'multipart-post', '~> 2.0', :require => false, :platforms => [:ruby, :jruby]`

File: httpi.gemspec [hunk 5]
```diff
@@ -19,11 +19,11 @@ Gem::Specification.new do |s|
...
   s.add_dependency 'socksify'
 
-  s.add_development_dependency 'rubyntlm', '~> 0.3.2'
-  s.add_development_dependency 'rake',     '~> 10.0'
-  s.add_development_dependency 'rspec',    '~> 2.14'
-  s.add_development_dependency 'mocha',    '~> 0.13'
-  s.add_development_dependency 'puma',     '~> 2.3.2'
+  s.add_development_dependency 'rubyntlm',    '~> 0.3.2'
+  s.add_development_dependency 'rake',        '~> 10.0'
+  s.add_development_dependency 'rspec',       '~> 3.4.0'
```
Comment (rogerleite) [inline httpi.gemspec]: Is there any reason specific to upgrade rspec here to 3.4.0?
I'm in favor to do this kind of change on another PR if you don't need anything specific from rspec 3.4.

File: spec/integration/support/application.rb [hunk 6]
```diff
@@ -1,6 +1,8 @@
 require 'rack'
 require "rack/builder"
 
+require "composite_io"
```
Comment (rogerleite) [inline spec/integration/support/application.rb]: I don't get this require. 😕

**Comment by rogerleite** [comment]
About the request body changes, i'll checkout your branch here to try to understand better. I would like to avoid changes on `net_http` adapter if we can.

Thanks! 

PS: Your plan is to use this adapter on Savon? Is there any issue or pull request related to link for?

File: spec/integration/support/application.rb (hunk 6 above)
Comment (keeguon) [inline spec/integration/support/application.rb]: Indeed it should be in the multipart spec, moved it.

File: httpi.gemspec (hunk 5 above)
Comment (keeguon) [inline httpi.gemspec]: Well, I was fiddling with gems and `rspec` got updated so I updated in the gemspec file as well but if this is too much of an issue I'll revert it to `~> 2.14`.

**Comment by keeguon** [comment]
@rogerleite yes, I'm trying to implement this: https://www.w3.org/TR/SOAP-attachments so I'll probably fiddle with httpi again today. As a matter of facts, shouldn't have rushed this PR but eh. As for the changes in Savon I'll do a PR when everything is working properly.
